src/gui/qgsrelationeditorwidget.cpp

```diff
@@ -636,8 +638,15 @@ void QgsRelationEditorWidget::updateUiSingleEdit()
       filters << filter.prepend( '(' ).append( ')' );
     }
 
-    QgsFeatureRequest nmRequest;
-    nmRequest.setFilterExpression( filters.join( QLatin1String( " OR " ) ) );
+    QString reducedExpression;
+    if ( QgsExpression::attemptReduceToInClause( filters, reducedExpression ) )
```

Comment:
I don't understand. Do we want to reduce to in clause beforehand here, or at eval time in prepareNode?

Reply:
well, they are independent: this change actually fixes the bug and speeds up this particular use case but this doesn't fix the general use case of expressions that contain OR chains that might be generated in other situations, so I decided to also address the general case at the expression level and that is done in qgsexpressionnodeimpl.cpp.